### PR TITLE
TextServerAdvanced: 2x performance improvement by removing redundant lookups

### PR DESCRIPTION
**Pull request by rune-scape**
- removes redundant font for size cache lookups
- also caches editor setting
- usies a faster hash method on the FontForSizeAdvanced cache (its the same function its just not converting the key to a variant before hashing)
- ShapedTextDataAdvanced::valid changed to an SafeFlag to avoid locking a mutex, as its read Very often

in total from my benchmark it takes ~39% of the time to render the same text in a Label (~45% in CodeEdit, the main script editor in godot)
all of these optimizations came from profiling a build using `target=editor optimize=speed_trace` they all gave at least a 10% improvement to speed at the time of profiling
now when profiled, the actual hotpath goes to freetype rendering and shaping and such

benchmark: [TextServerTest.zip](https://github.com/user-attachments/files/15508175/TextServerTest.zip)
benchmark compile flags: `scons target=template_release production=yes optimize=speed lto=full tests=yes deprecated=no scu_build=yes module_text_server_adv_enabled=yes module_text_server_fb_enabled=no`
benchmark run command: `./godot/bin/godot.linuxbsd.template_release.x86_64 --disable-vsync ./TextServerTest/project.godot`

results before:
```
Godot Engine v4.3.beta.custom_build.6b38be699 (2024-05-30 20:38:39 UTC) - https://godotengine.org
Vulkan 1.3.277 - Forward+ - Using Device #0: NVIDIA - NVIDIA GeForce GTX 1080 Ti

18.1370287230521ms avg draw time (99 samples)
18.1469531974407ms avg draw time (198 samples)
18.1223431018868ms avg draw time (297 samples)
18.159361198695ms avg draw time (396 samples)
18.1944639995845ms avg draw time (495 samples)
18.1336226286712ms avg draw time (594 samples)
18.14051692792ms avg draw time (693 samples)
18.1234557219226ms avg draw time (792 samples)
18.0797271856957ms avg draw time (891 samples)
18.0556610377148ms avg draw time (990 samples)
18.0405357542993ms avg draw time (1089 samples)
```
after:
```
Godot Engine v4.3.beta.custom_build.02f8b6443 (2024-05-30 22:54:24 UTC) - https://godotengine.org
Vulkan 1.3.277 - Forward+ - Using Device #0: NVIDIA - NVIDIA GeForce GTX 1080 Ti

7.03977334378946ms avg draw time (99 samples)
7.00848030321526ms avg draw time (198 samples)
7.01666921878905ms avg draw time (297 samples)
7.02174504597982ms avg draw time (396 samples)
7.03076449307528ms avg draw time (495 samples)
7.036976942711ms avg draw time (594 samples)
7.03303741686272ms avg draw time (693 samples)
7.02608956231011ms avg draw time (792 samples)
7.01870688150494ms avg draw time (891 samples)
7.01467798213766ms avg draw time (990 samples)
7.0109612576123ms avg draw time (1089 samples)
```

the last line of each is the most important, its the average of all frame times (minus some longer startup frames)

the same improvements can be made to TextServerFallback

### CODE REVIEW

File: modules/text_server_adv/text_server_adv.h [hunk 1]
```diff
@@ -487,7 +491,7 @@ class TextServerAdvanced : public TextServerExtension {
 		/* Shaped data */
 		TextServer::Direction para_direction = DIRECTION_LTR; // Detected text direction.
 		int base_para_direction = UBIDI_DEFAULT_LTR;
-		bool valid = false; // String is shaped.
+		SafeFlag valid = SafeFlag(false); // String is shaped.
```
Comment (AThousandShips) [inline modules/text_server_adv/text_server_adv.h:494]:
```suggestion
		SafeFlag valid; // String is shaped.
```
In this case it's now unnecessary to assign

File: modules/text_server_adv/text_server_adv.h [hunk 2]
```diff
@@ -151,6 +152,9 @@ class TextServerAdvanced : public TextServerExtension {
 	HashMap<StringName, int32_t> feature_sets;
 	HashMap<int32_t, FeatureInfo> feature_sets_inv;
 
+	SafeNumeric<TextServer::FontLCDSubpixelLayout> lcd_subpixel_layout = SafeNumeric<TextServer::FontLCDSubpixelLayout>(TextServer::FontLCDSubpixelLayout::FONT_LCD_SUBPIXEL_LAYOUT_NONE);
```
Comment (AThousandShips) [inline modules/text_server_adv/text_server_adv.h:155]:
```suggestion
	SafeNumeric<TextServer::FontLCDSubpixelLayout> lcd_subpixel_layout{ TextServer::FontLCDSubpixelLayout::FONT_LCD_SUBPIXEL_LAYOUT_NONE };
```
Seems to be preferred syntax elsewhere

Comment (rune-scape) [reply modules/text_server_adv/text_server_adv.h]: i see,, i kinda assume c++11 syntax will be rejected if i use it, bc auto and lambdas aren't allowed, also srry for the commit spam , couldn't figure out how to install the precommit git hooks

Comment (AThousandShips) [reply modules/text_server_adv/text_server_adv.h]: You can also just use parentheses, but haven't seen assignment like that

Comment (rune-scape) [reply modules/text_server_adv/text_server_adv.h]: using parenthesis makes the c++ compiler assume its a function and `SafeNumeric` has only an explicit constructor,, so these are the only ways im aware of

**Review by bruvzg — APPROVED** [review]
The only thing I would change is returning value from the `_ensure_cache_for_size` as pointer instead of reference, to avoid unnecessary local variable which are unused in most cases.
```diff
-	FontForSizeAdvanced *ffsd = nullptr;
-	ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, size, ffsd), 0);
+	ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, size, nullptr), 0);
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Needs a rebase, which should show a compilation error (tested locally):
```
In file included from ./core/string/char_utils.h:34,
                 from ./core/string/ustring.h:36,
                 from modules/text_server_adv/script_iterator.h:46,
                 from modules/text_server_adv/text_server_adv.h:39,
                 from modules/text_server_adv/text_server_adv.cpp:31:
modules/text_server_adv/text_server_adv.cpp: In member function 'virtual PackedInt32Array TextServerAdvanced::_font_get_supported_glyphs(const RID&) const':
modules/text_server_adv/text_server_adv.cpp:3603: error: no matching function for call to 'TextServerAdvanced::_ensure_cache_for_size(TextServerAdvanced::FontAdvanced*&, Vector2i) const'
 3603 |                 ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, fd->msdf ? Vector2i(fd->msdf_source_size, 0) : Vector2i(16, 0)), PackedInt32Array());
./core/typedefs.h:272:41: note: in definition of macro 'unlikely'
  272 | #define unlikely(x) __builtin_expect(!!(x), 0)
      |                                         ^
modules/text_server_adv/text_server_adv.cpp:3603: note: in expansion of macro 'ERR_FAIL_COND_V'
 3603 |                 ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, fd->msdf ? Vector2i(fd->msdf_source_size, 0) : Vector2i(16, 0)), PackedInt32Array());
modules/text_server_adv/text_server_adv.cpp:1360: note: candidate: 'bool TextServerAdvanced::_ensure_cache_for_size(FontAdvanced*, const Vector2i&, FontForSizeAdvanced*&) const'
 1360 | _FORCE_INLINE_ bool TextServerAdvanced::_ensure_cache_for_size(FontAdvanced *p_font_data, const Vector2i &p_size, FontForSizeAdvanced *&r_cache_for_size) const {
modules/text_server_adv/text_server_adv.cpp:1360: note:   candidate expects 3 arguments, 2 provided
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**